whisper.cpp

```diff
@@ -2432,34 +2431,43 @@ static void log_mel_spectrogram_worker_thread(int ith, const std::vector<float>
                                               const whisper_filters &filters, bool speed_up, whisper_mel &mel) {
     std::vector<float> fft_in(fft_size, 0.0);
     std::vector<float> fft_out(2 * fft_size);
+    // Is using 32-bit float to calculate log_mel appropriate?
+    // 32-bit float has about 7 digits of precision, but minimum value of log_mel is 1e-10.
     int n_fft = 1 + (speed_up ? fft_size / 4 : fft_size / 2);
+    int i = ith;
 
-    for (int i = ith; i < mel.n_len; i += n_threads) {
+    // Calculate FFT only when fft_in are not all zero
+    for (; i < std::min((n_samples / fft_step) + 1, mel.n_len); i += n_threads) {
         const int offset = i * fft_step;
 
-        // apply Hanning window
-        for (int j = 0; j < fft_size; j++) {
-            if (offset + j < n_samples) {
-                fft_in[j] = hann[j] * samples[offset + j];
-            } else {
-                fft_in[j] = 0.0;
-            }
+        // apply Hanning window (~10% faster)
+        for (int j = 0; j < std::min(fft_size, n_samples - offset); j++) {
+            fft_in[j] = hann[j] * samples[offset + j];
+        }
+        // fill the rest with zeros
+        if (n_samples - offset < fft_size) {
```

Comment:
Isn't doing a zero fill here problematic?  Let's say you've got an `fft_size` of `256`, and `n_samples - offset` happens to come out to `128`.  That means you're creating a cliff edge between `fft_in[127]` and `fft_in[128]`.  The Hanning window is at its maximum value there, so you get no smoothing at all, it's going to be all artefacty.

Reply:
Good question! So, I am writing a more detailed code analysis note, comparing the differences in the methods of generating log mel spectrograms between OpenAI's whisper and whisper.cpp. I have already completed part of it, and I will publish this finished portion shortly.

Reply:
I've got a sneaking suspicion it might not be easy to get this right without doing some rework on the audio pipeline.  Ideally you'd want to be saving unprocessed samples for the next round so you're always working with a full buffer.